tests/Unit/CssInlinerTest.php

```diff
@@ -54,364 +50,11 @@ private function buildDebugSubject($html)
-     */
-    public function contentWithoutHeadTagDataProvider()
-    {
-        return [
-            'doctype only' => ['<!DOCTYPE html>'],
-            'body content only' => ['<p>Hello</p>'],
-            'BODY element' => ['<body></body>'],
-        ];
-    }
-
-    /**
-     * @test
-     *
-     * @param string $html
-     *
-     * @dataProvider contentWithoutHeadTagDataProvider
-     */
-    public function renderAddsMissingHeadTag($html)
-    {
-        $subject = $this->buildDebugSubject($html);
-
-        $result = $subject->render();
-
-        self::assertContains('<head>', $result);
-    }
-
-    /**
-     * @return string[][]
-     */
-    public function contentWithoutBodyTagDataProvider()
-    {
-        return [
-            'doctype only' => ['<!DOCTYPE html>'],
-            'HEAD element' => ['<head></head>'],
-            'body content only' => ['<p>Hello</p>'],
-        ];
-    }
-
-    /**
-     * @test
-     *
-     * @param string $html
-     *
-     * @dataProvider contentWithoutBodyTagDataProvider
-     */
-    public function renderAddsMissingBodyTag($html)
-    {
-        $subject = $this->buildDebugSubject($html);
-
-        $result = $subject->render();
-
-        self::assertContains('<body>', $result);
-    }
-
-    /**
-     * @test
-     */
-    public function renderPutsMissingBodyElementAroundBodyContent()
-    {
-        $subject = $this->buildDebugSubject('<p>Hello</p>');
-
-        $result = $subject->render();
-
-        self::assertContains('<body><p>Hello</p></body>', $result);
-    }
-
-    /**
-     * @return string[][]
-     */
-    public function specialCharactersDataProvider()
-    {
-        return [
-            'template markers with dollar signs & square brackets' => ['$[USER:NAME]$'],
-            'UTF-8 umlauts' => ['Küss die Hand, schöne Frau. イリノイ州シカゴにて、アイルランド系の家庭に、'],
-            'HTML entities' => ['a &amp; b &gt; c'],
-            'curly braces' => ['{Happy new year!}'],
-        ];
-    }
-
-    /**
-     * @test
-     *
-     * @param string $codeNotToBeChanged
-     *
-     * @dataProvider specialCharactersDataProvider
-     */
-    public function renderKeepsSpecialCharactersInTextNodes($codeNotToBeChanged)
-    {
-        $html = '<html><p>' . $codeNotToBeChanged . '</p></html>';
-        $subject = $this->buildDebugSubject($html);
-
-        $result = $subject->render();
-
-        self::assertContains($codeNotToBeChanged, $result);
-    }
-
-    /**
-     * @test
-     */
-    public function renderAddsMissingHtml5DocumentType()
-    {
-        $subject = $this->buildDebugSubject('<html><h1>foo</h1></html>');
-
-        $result = $subject->render();
-
-        self::assertContains('<!DOCTYPE html>', $result);
-    }
-
-    /**
-     * @test
-     *
-     * @param string $codeNotToBeChanged
-     *
-     * @dataProvider specialCharactersDataProvider
-     */
-    public function renderBodyContentKeepsSpecialCharactersInTextNodes($codeNotToBeChanged)
-    {
```

Comment:
I have now also spotted that `AbstractHtmlProcessorTest` does not have this test specifically for `renderBodyContent` (it has just the more general `keepsSpecialCharactersInTextNodes`) - should it (e.g. perhaps the procedure to extract the body content could affect special characters)?

Reply:
To be sure, yes. I've added the test and will repush in a second.

Reply:
Looks good,